activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -104,6 +104,24 @@ def take!
       take || raise_record_not_found_exception!
     end
 
+    # Finds the only matching record. Raises ActiveRecord::RecordNotFound if no
+    # record is found. Raises ActiveRecord::MultipleRecordsFound if more than
+    # one record is found.
+    #
+    #   Product.where(["price = %?", price]).only!
+    def only!
+      results = all
+      if results.length == 0
+        raise_record_not_found_exception!
+      elsif results.length > 1
+        error = +"Wanted only one #{name}, returned #{results.length}"
+        conditions = " [#{arel.where_sql(klass)}]" unless where_clause.empty?
+        error << " with#{conditions}" if conditions
+        raise ActiveRecord::MultipleRecordsFound.new(error, name, primary_key)
+      end
+      results.first
+    end
```

Comment:
Feels like we'd be doing a fair amount of queries here, and we could try sidestepping that by loading 2 records. This should be fine since the second record should pretty much always be nil in the base case.

```suggestion
def only!
  found, undesired = first(2)
  raise_record_not_found_exception! if found.nil?

  if undesired
    error = +"Wanted only one #{name}"
    conditions = " [#{arel.where_sql(klass)}]" unless where_clause.empty?
    error << " with#{conditions}" if conditions
    raise ActiveRecord::MultipleRecordsFound.new(error, name, primary_key)
  end

  found
end
```